### DISCUSSION TITLE
[Section 3.14 - Interpolação, locals e outputs] - Problema na validação do output

### DISCUSSION THREAD

**Discussion by bernardesleao** [chgasparoto/curso-aws-com-terraform#144] [Geral]
Olá, estou com dificuldades em validar a configuração dos arquivos terraform, referente ao item Outputs.
Para meus estudos, criei uma estrutura de diretórios separados em meu Visual Studio Code para cada módulo desta seção (um para tratar do assunto Interpolação, outro para locals, etc.), criando arquivos individuais de cada aula.

No entanto, mesmo seguindo as orientações passadas no vídeo, reconstruindo os arquivos individualmente, recebo a mensagem de erro abaixo, quando executo o comando "terraform plan -out="tfplan.out""
O que sugere o erro?

Aqui está meu código do arquivo s3.tf

``` hcl
resource "aws_s3_bucket" "this" {
  bucket = "${random_pet.bucket.id}-${var.aws_environment}"
  tags   = local.common_tags
}

resource "aws_s3_bucket_object" "this" {
  bucket = aws_s3_bucket.this.bucket
  key    = "config/${local.ip_filepath}"
  source = local.ip_filepath
  etag   = filemd5(local.ip_filepath)
}
```

Aqui está meu código do arquivo variables.tf

``` hcl
variable "aws_region" {
  type        = string
  description = "value"
  default     = "us-east-1"
}

variable "aws_environment" {
  type        = string
  description = "value"
  default     = "development"
}
```

Aqui está meu código do arquivo outputs.tf

``` hcl
output "bucket_name" {
  value = aws_s3_bucket.this.bucket
}

output "arn" {
  value = aws_s3_bucket.this.arn
}

output "domain_name" {
  value = aws_s3_bucket.this.domain_name
}

output "ips_file_path" {
  value = "${aws_s3_bucket.this.bucket}/${aws_s3_bucket_object.this.key}"
}
```

Aqui está meu código do arquivo locals.tf

``` hcl
locals {
  ip_filepath = "ips.json"

  common_tags = {
    Name        = "Meu bucket S3"
    Environment = "var.aws_environment"
    ManagedBy   = "terraform"
    Owner       = "Amancio"
  }
}
```

Aqui está meu código do arquivo main.tf

``` hcl
terraform {
  required_version = "1.5.2"
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "5.6.2"
    }
    random = {
      source  = "hashicorp/random"
      version = "3.5.1"
    }
  }
}

provider "aws" {
  region = var.aws_region
}
```

E aqui está a mensagem que recebo, após executar o comando terraform plan:

``` hcl
Error: Reference to undeclared resource
│
│   on s3.tf line 2, in resource "aws_s3_bucket" "this":
│    2:   bucket = "${random_pet.bucket.id}-${var.aws_environment}"
│
│ A managed resource "random_pet" "bucket" has not been declared in the root module.
```

**Comment by chgasparoto** [comment]
@bernardesleao pelo que vi no seu código, o erro está no nome do seu bucket. Tente alterar o `bucket = "${random_pet.bucket.id}-${var.aws_environment}"`por uma string aleatória ou defina um bloco de `random_pet` para o recurso de bucket conseguir setar o nome corretamente.

Exemplo:
https://github.com/chgasparoto/curso-aws-com-terraform/blob/master/01-terraform-basico/04-interpolacao-locals-outputs/random.tf
